include/seqan3/range/container/small_string.hpp

```diff
@@ -0,0 +1,265 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \author Svenja Mehringer <svenja.mehringer AT fu-berlin.de>
+ * \brief A constexpr string implementation to manipulate string literals at compile time.
+ */
+
+#pragma once
+
+#include <seqan3/range/container/small_vector.hpp>
+
+namespace seqan3
+{
+
+/*!\brief Implements a small string that can be used for compile time computations.
+ * \ingroup container
+ * \tparam N The size of the small string.
+ *
+ * This class provides a string type that can be constructed, evaluated and operated on at compile time.
+ * The stored string can be accessed as either a std::string or a c-style string through the respective member functions.
+ * All operations are constexpr except the conversion functions.
+ *
+ * Internally the string stores an array of size `N+1` containing the `0`-byte to allow conversion to
+ * c-style strings. This also means that when creating an instance of seqan3::small_string from a string literal
+ * containing a `0`-byte the size() of this small_string instance is one less than that of the literal.
+ *
+ * Example:
+ *
+ * \include test/snippet/range/container/small_string.cpp
+ *
+ */
+template <size_t capacity_>
+class small_string : public small_vector<char, capacity_ + 1>
+{
+    //!\brief The parent class.
+    using base_t = small_vector<char, capacity_ + 1>;
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    using typename base_t::value_type;
+    using typename base_t::reference;
+    using typename base_t::const_reference;
+    using typename base_t::iterator;
+    using typename base_t::const_iterator;
+    using typename base_t::difference_type;
+    using typename base_t::size_type;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    //!\brief Inherit the parent class constructors.
+    using base_t::base_t;
+    //!\brief Inherit the parent class member function assign.
+    using base_t::assign;
+
+    //!\brief The default constructor initializes the underlying data member with '\0's.
+    constexpr small_string() noexcept : base_t{}
+    {
+        for (size_t i = 0; i < capacity_ + 1; ++i)
+            this->data_[i] = '\0';
+    }
```

Comment:
why doesn't `base_t{}` do the zeroing?

Reply:
why should it? the `small_vector` does never access the element one past the end and does not have the `.c_str()` functionality. It does brace initialization of the underlying array on default construction... I am not sure if this default initializes the elements too.

Reply:
> On second thought, It does brace initialization of the underlying array on default construction... I am not sure if this default initializes the elements too.

It does, that's what I meant!

Reply:
I removed the constructor